src/FffGcodeWriter.cpp

```diff
@@ -1455,168 +1456,275 @@ bool FffGcodeWriter::processSkinAndPerimeterGaps(const SliceDataStorage& storage
+        if (gcode_layer.getLayerNr() > 0)
         {
-            perimeter_gaps = &perimeter_gaps_recomputed;
-            perimeter_gaps_recomputed = concentric_perimeter_gaps.unionPolygons(skin_part.perimeter_gaps);
+            bridge = bridgeAngle(skin_part.outline, &mesh.layers[gcode_layer.getLayerNr() - 1]);
         }
-        else if (extruder_nr == wall_0_extruder_nr)
+        if (bridge > -1)
         {
-            perimeter_gaps = &skin_part.perimeter_gaps;
+            pattern = EFillMethod::LINES; // force lines pattern when bridging
+            skin_angle = bridge;
         }
-        else
+
+        // calculate polygons and lines
+        constexpr int extra_infill_shift = 0;
+        constexpr coord_t offset_from_inner_skin_infill = 0;
+        const coord_t skin_overlap = mesh.getSettingInMicrons("skin_overlap_mm");
+        Polygons* perimeter_gaps_output = (generate_perimeter_gaps)? &concentric_perimeter_gaps : nullptr;
+        Infill infill_comp(pattern, skin_part.inner_infill, offset_from_inner_skin_infill, skin_line_width, skin_line_width, skin_overlap, skin_angle, gcode_layer.z, extra_infill_shift, perimeter_gaps_output);
+        infill_comp.generate(skin_polygons, skin_lines);
+    }
+
+    // add skin itself!
+    if (process_skin_infill && (skin_polygons.size() > 0 || skin_lines.size() > 0))
+    {
+        added_something = true;
+        setExtruder_addPrime(storage, gcode_layer, extruder_nr);
+        gcode_layer.setIsInside(true); // going to print stuff inside print object
+        gcode_layer.addPolygonsByOptimizer(skin_polygons, mesh_config.skin_config);
+
+        if (pattern == EFillMethod::GRID || pattern == EFillMethod::LINES || pattern == EFillMethod::TRIANGLES || pattern == EFillMethod::CUBIC || pattern == EFillMethod::TETRAHEDRAL || pattern == EFillMethod::CUBICSUBDIV)
         {
-            perimeter_gaps = &concentric_perimeter_gaps;
+            gcode_layer.addLinesByOptimizer(skin_lines, mesh_config.skin_config, SpaceFillType::Lines, mesh.getSettingInMicrons("infill_wipe_dist"));
         }
-        Polygons gap_polygons; // will remain empty
-        Polygons gap_lines;
-        int offset = 0;
-        Infill infill_comp(EFillMethod::LINES, *perimeter_gaps, offset, perimeter_gaps_line_width, perimeter_gaps_line_width, skin_overlap, skin_angle, z, extra_infill_shift);
-        infill_comp.generate(gap_polygons, gap_lines);
-        if (gap_lines.size() > 0)
+        else
         {
-            added_something = true;
-            setExtruder_addPrime(storage, gcode_layer, layer_nr, extruder_nr);
-            gcode_layer.setIsInside(true); // going to print stuff inside print object
-            gcode_layer.addLinesByOptimizer(gap_lines, mesh_config.perimeter_gap_config, SpaceFillType::Lines);
+            gcode_layer.addLinesByOptimizer(skin_lines, mesh_config.skin_config, (pattern == EFillMethod::ZIG_ZAG)? SpaceFillType::PolyLines : SpaceFillType::Lines);
         }
     }
```

Comment:
Lots of this code is duplicate from the `processSkinPartRoofingInfillGeneratePerimeterGaps` function. Once you've broken these functions up, please remember to re-use the parts where possible.